### PR TITLE
BTSSS - Allow filtering by date range

### PR DESCRIPTION
**Pull request by hinzed1127**
## Summary

- Adds ability to filter claims by pre-generated date ranges
- Works in tandem with status filtering
- Updates checkbox UX (no status filters checked = show all claims)

## Related issue(s)

- https://app.zenhub.com/workspaces/beneficiary-travel-btsss-65147e21930cd900223d8e64/issues/gh/department-of-veterans-affairs/va.gov-team/83450

## Testing done

- Thorough unit testing of the date range creation utility 
- Cypress tests (and what unit testing RTL will let us do w/ web components)

## Screenshots

[Loom demo](https://www.loom.com/share/bdb43d8fbc94486280a376d1a4355eed?sid=dcc59504-45aa-45ab-b289-18465c14a163)
Screenshot 2024-06-25 at 4 18 40 PM

_Note: This field is mandatory for UI changes (non-component work should NOT have screenshots)._

|         | Before | After |
| ------- | ------ | ----- |
| Mobile  |        |       |
| Desktop |        |       |

## What areas of the site does it impact?

Travel Pay

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [x] Documentation has been updated ([link to documentation](#) \*if necessary)
- [x] Screenshot of the developed feature is added
- [x] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [x] Browser console contains no warnings or errors.
- [x] Events are being sent to the appropriate logging solution
- [x] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [x] Did you login to a local build and verify all authenticated routes work as expected with a test user

### :warning: Team Sites (only applies to modifications made to the VA.gov header) :warning:

- [ ] The vets-website header does not contain any web-components
- [ ] I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#local-dev) to test the injected header scenario
- [ ] I reached out in the `#sitewide-public-websites` Slack channel for questions

## Requested Feedback

(OPTIONAL) _What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: src/applications/travel-pay/components/TravelPayClaimFilters.jsx [hunk 1]
```diff
@@ -19,11 +22,30 @@ export default function TravelPayFilters(props) {
         header="Filter travel claims"
       >
         <div className="filter-your-results">
-          <TravelPayStatusFilters
```
Comment (athifw) [inline src/applications/travel-pay/components/TravelPayClaimFilters.jsx:22]: our idea with this was that we'd have `TravelPayStatusFilters` be a parent component to hold both the status checkboxes component and the date filter component. What do you think of making the date filter its own component, and then incorporating it into `TravelPayStatusFilters`?

Comment (kjduensing) [reply src/applications/travel-pay/components/TravelPayClaimFilters.jsx:22]: I second this - I didn't catch it in the status PR, but we should rename `<TravelPayStatusFilters />` to be more general: `<TravelPayFilterConterainer />` or something

File: src/applications/travel-pay/util/dates.js [hunk 2]
```diff
@@ -7,3 +37,63 @@ export function formatDateTime(datetimeString) {
...
+    );
+
+    dateRanges.push({
+      label: `${startMonth.label} ${quarterYear} - ${
+        endMonth.label
+      } ${quarterYear}`,
+      start: new Date(quarterYear, startMonth.value - 1, 1),
+      end: endOfMonth(new Date(quarterYear, endMonth.value - 1)),
+    });
+  }
```
Comment (kjduensing) [inline src/applications/travel-pay/util/dates.js:86]: Thinking out loud here - it seems like we could increase the clarity of `getDateFilters()` with something like:

```javascript
const dateRangeLabel = (start, end) => {
  return `${format(quarterStart, 'MMM yyyy')} - ${format(quarterEnd, 'MMM yyyy'}`
}

dateRanges.push({
  label: dateRangeLabel(quarterStart, quarterEnd),
  start: quarterStart,
  end: quarterEnd,
});
```

Comment (kjduensing) [reply src/applications/travel-pay/util/dates.js:86]: What are your thoughts? I think this would remove the need for a lot of intermediary variables and structure, like the lookup table defined above could go away with this.

**Comment by kjduensing** [comment]
@hinzed1127 I see from the screenshot that the status filters are "screaming snake case". I can't see anything in the code that would change the status filters from human readable ("In process") to all caps underscore (IN_PROCESS). Do you know why the status filter options in the screenshot are different from the status filter options in staging?

File: src/applications/travel-pay/components/TravelPayClaimFilters.jsx (hunk 1 above)
Comment (hinzed1127) [reply src/applications/travel-pay/components/TravelPayClaimFilters.jsx:22]: good call; keeps things more composable and clean, sets us up better if/when we add more to the filtering.

File: src/applications/travel-pay/util/dates.js (hunk 2 above)
Comment (hinzed1127) [reply src/applications/travel-pay/util/dates.js]: Good call. I was definitely overusing variables a bit. Cleaned it up and it's a lot more straightforward now

**Comment by hinzed1127** [comment]
@kjduensing I can't remember when/why the claim status mapping got removed on the FE, but I added it back in. Previously was just in `TravelClaimCard`, but now that we've got them in filtering, too, I created ` util/constants.js` to have those in a shared place

**Comment by kjduensing** [comment]
@hinzed112 we talked about this but I dunno that we ever look at the code - the appointments team does this same filtering method - [anything in this code that could help us](https://github.com/department-of-veterans-affairs/vets-website/blob/cd4de813eb4a55c2465f8138c92e991204d62c80/src/applications/vaos/appointment-list/components/PastAppointmentsList)?

**Comment by hinzed1127** [comment]
> [anything in this code that could help us](https://github.com/department-of-veterans-affairs/vets-website/blob/cd4de813eb4a55c2465f8138c92e991204d62c80/src/applications/vaos/appointment-list/components/PastAppointmentsList)?

Good sentiment, and I definitely referenced that when working on this. The main pain points that came up related to that:

1. Appointments is using `moment.js` and `vets-website` is trying to shift to `date-fns` for any new development. This was the same issue with the previous date stuff we added here.
2. There's enough business logic tied to appointments that abstracting it seems like more trouble than its worth for now

I think it's worth keeping an eye on as we add to it, but for now, the juice probably isn't worth the squeeze.

File: src/applications/travel-pay/containers/TravelPayStatusApp.jsx [hunk 3]
```diff
@@ -69,24 +88,40 @@ export default function App({ children }) {
   }
 
   const resetSearch = () => {
-    setAppliedStatuses(statusesToFilterBy);
-    setCheckedStatuses(statusesToFilterBy);
+    setAppliedStatusFilters([]);
+    setCheckedStatusFilters([]);
```
Comment (athifw) [inline src/applications/travel-pay/containers/TravelPayStatusApp.jsx:92]: it looks like this resets and initiates the search with checkboxes that aren't checked. Is this expected behavior?

Comment (hinzed1127) [reply src/applications/travel-pay/containers/TravelPayStatusApp.jsx:92]: Yeah. Sorry, I should have updated the ticket somewhere with that. [Here's some back-and-forth with Micah regarding that](https://adhoc.slack.com/archives/C05R710CVMH/p1719349406346699?thread_ts=1719244041.433189&cid=C05R710CVMH). 

TL;DR
> No status filters checked = show all claims

File: src/applications/travel-pay/containers/TravelPayStatusApp.jsx [hunk 4]
```diff
@@ -69,24 +88,40 @@ export default function App({ children }) {
...
+        checkedStatusFilters.filter(
+          statusFilter => statusFilter !== statusName,
+        ),
       );
     }
   };
 
+  const onDateFilterChange = e => {
+    setSelectedDateFilter(e.target.value);
+    setCurrentPage(1);
```
Comment (athifw) [inline src/applications/travel-pay/containers/TravelPayStatusApp.jsx:117]: this sets the current page even when the filters haven't been applied. Shouldn't we only reset the current page after a search is submitted?

Comment (hinzed1127) [reply src/applications/travel-pay/containers/TravelPayStatusApp.jsx]: Good catch, meant for that to be in `applyFilters`, not `onDateFilterChange`

**Review by athifw — APPROVED** [review]

**Review by IGallupSoCo — APPROVED** [review]
LGTM
